core/project.go

```diff
@@ -0,0 +1,403 @@
+package core
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+
+	"github.com/dagger/dagger/router"
+	"github.com/dagger/graphql"
+	"github.com/dagger/graphql/language/ast"
+	"github.com/dagger/graphql/language/parser"
+	"github.com/moby/buildkit/client/llb"
+	bkgw "github.com/moby/buildkit/frontend/gateway/client"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/vektah/gqlparser/v2"
+	gqlparserast "github.com/vektah/gqlparser/v2/ast"
+)
+
+const (
+	schemaPath     = "/schema.graphql"
+	entrypointPath = "/entrypoint"
+
+	tmpMountPath = "/tmp"
+
+	inputMountPath = "/inputs"
+	inputFile      = "/dagger.json"
+
+	outputMountPath = "/outputs"
+	outputFile      = "/dagger.json"
+)
+
+type ProjectID string
+
+func (id ProjectID) String() string {
+	return string(id)
+}
+
+func (id ProjectID) ToProject() (*Project, error) {
+	var project Project
+	if id == "" {
+		return &project, nil
+	}
+	if err := decodeID(&project, id); err != nil {
+		return nil, err
+	}
+	return &project, nil
+}
+
+type ProjectCommandID string
+
+func (id ProjectCommandID) String() string {
+	return string(id)
+}
+
+func (id ProjectCommandID) ToProjectCommand() (*ProjectCommand, error) {
+	var projectCommand ProjectCommand
+	if id == "" {
+		return &projectCommand, nil
+	}
+	if err := decodeID(&projectCommand, id); err != nil {
+		return nil, err
+	}
+	return &projectCommand, nil
+}
+
+type Project struct {
+	Directory  *Directory     `json:"directory"`
+	ConfigPath string         `json:"configPath"`
+	Config     ProjectConfig  `json:"config"`
+	Schema     string         `json:"schema"`
```

Comment:
Note to self: this field stores a generated GraphQL schema

Reply:
Will add comments to each field before merging